### PR TITLE
GSoC 2022 Manas Sivakumar : Third and Fourth Week

### PR DESCRIPTION
**Pull request by Manas23601**
Fixes # .

Changes proposed in this pull request:
-
-
-

@pgRouting/admins

### CODE REVIEW

**Comment by Manas23601** [comment]
for knapsack_driver.h in include/or-tools do I have to make my own c_type structures as well? 
I have a few question regarding the implementation:
1) the knapsack function takes 3 arguments:
- bin capacity
- weights
- values
For now I have set the weights and values to be SQL  text so that users can fetch data from tables. I believe no such tables related to knapsack already exist. I think this means we don't any structures that have access to other data points like coordinates, vehichle_id, etc. Now this raises two questions to me. 1) Do I really need my own structs and 2) Should I change the SQL data inputs such that weights and values are arrays of integers with 'N' items.

**Comment by krashish8** [comment]
Q: Why do you want to access other data points like coordinates, vehicle_id for your knapsack function?

For the structure of your function, there are following possibilities:
- Both weights and values are arrays of integers.
- A table has 2 columns [weight, value] and the data is fetched from an SQL query.
- You have 2 separate tables for weight and value, and you run 2 separate SQL queries to fetch the data.

I think the third approach is not a good one, because why run an SQL query for a table having a single column?

You can go with either the 1st or the 2nd option. The 1st approach is intuitive because say a user has a table with all the item details, then for the knapsack function, the user can fetch two columns from that table. The 2nd approach is also okay, because any way you just require weight and value for the knapsack, and don't need any additional data, so both of these values can be fed in the form of an array. Though in this approach, you need to do an extra check that the length of weight and value arrays are the same.

Yes, if such a struct does not exist, then you have to create it.

File: sql/or_tools/CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,6 @@
 SET(LOCAL_FILES
-  _knapsack_0-1.sql
+  _knapsack.sql
+  _knapsack.sql
```
Comment (cvvergara) [inline sql/or_tools/CMakeLists.txt:3]: The user facing function file goes without underscore.
you are including the same file twice

**Comment by Manas23601** [comment]
> Q: Why do you want to access other data points like coordinates, vehicle_id for your knapsack function?
> 
> For the structure of your function, there are following possibilities:
> 
> * Both weights and values are arrays of integers.
> * A table has 2 columns [weight, value] and the data is fetched from an SQL query.
> * You have 2 separate tables for weight and value, and you run 2 separate SQL queries to fetch the data.
> 
> I think the third approach is not a good one, because why run an SQL query for a table having a single column?
> 
> You can go with either the 1st or the 2nd option. The 1st approach is intuitive because say a user has a table with all the item details, then for the knapsack function, the user can fetch two columns from that table. The 2nd approach is also okay, because any way you just require weight and value for the knapsack, and don't need any additional data, so both of these values can be fed in the form of an array. Though in this approach, you need to do an extra check that the length of weight and value arrays are the same.
> 
> Yes, if such a struct does not exist, then you have to create it.

My question was not very clear. I updated it now. But I think you answered my question. Is there any need for a table with additional data columns other than weights and values ? I assume if people want to use the or-tools knapsack implementation they will be having a huge number of items so asking them to enter it as parameters would be difficult

**Pull request merged**